From 711ad583a21584cf7720e2eef5b253c83b4451d8 Mon Sep 17 00:00:00 2001
From: Yuchou Gan <yuchou.gan@nxp.com>
Date: Sat, 26 May 2018 00:17:55 +0800
Subject: [PATCH 3883/5242] MLK-18398 gpu: imx: imx8_dprc: dpu-blit: Wait the
 dprc idle before disable it

commit  0783888d91d5e5373855fcdeb9ded7b65c2dd615 from
https://source.codeaurora.org/external/imx/linux-imx.git

For linear and tiled mixed blit, wait the dprc idle
before disable it in linear blit.

Signed-off-by: Yuchou Gan <yuchou.gan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu-blit/dpu-blit.c |   16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/imx/dpu-blit/dpu-blit.c b/drivers/gpu/imx/dpu-blit/dpu-blit.c
index ef7162a..7df60e0 100644
--- a/drivers/gpu/imx/dpu-blit/dpu-blit.c
+++ b/drivers/gpu/imx/dpu-blit/dpu-blit.c
@@ -110,19 +110,29 @@ void dpu_be_configure_prefetch(struct dpu_bliteng *dpu_be,
 	static bool start = true;
 	static bool need_handle_start;
 	struct dprc *dprc;
+	static bool tiled_work_unfinished = true;
 
 	/* Enable DPR, dprc1 is connected to plane0 */
 	dprc = dpu_be->dprc[1];
 
+	/*
+	 * Waiting for the previous tiled command finished
+	 * before disable the dpr.
+	 */
+	if (tiled_work_unfinished) {
+		dpu_be_wait(dpu_be);
+		dpu_cs_wait_idle(dpu_be);
+		udelay(10);
+		tiled_work_unfinished = false;
+	}
+
 	if (baddr == 0x0) {
 		dprc_disable(dprc);
 		start = true;
 		return;
 	}
 
-	dpu_be_wait(dpu_be);
-	dpu_cs_wait_idle(dpu_be);
-	udelay(10);
+	tiled_work_unfinished = true;
 
 	if (need_handle_start) {
 		dprc_first_frame_handle(dprc);
-- 
1.7.9.5

